### ISSUE TITLE
API Reference documentation has tragic visual structure

### ISSUE THREAD

**Issue by teodorkostov** [numpy/numpy#19420]
## Documentation

[API Reference](https://numpy.org/doc/stable/reference/index.html)

The only way the this can be used currently is by searching in a third party search engine. If your users have to use your search to get information on your site - this is bad. If your users have to consistently use another search engine (e.g. Google) to get the page of your site they are looking for - that's even worse.

The point of this issue is not to just criticize the current visual structure of the API Reference. The idea is to provoke a conversation and redesign of the API Reference documentation visuals so that the following is achieved:

* All of the framework functionality is efficiently presented and described on a single page
* All of the framework functionality is efficiently searchable from within a single page (Ctrl+F)
* The detailed structure of the functionality and the documentation is not a burden but a enhancement for the user experience

### Poor documentation visual structure for API Reference

[This](https://lodash.com/docs/4.17.15) looks like an API Reference. [This](https://numpy.org/doc/stable/reference/index.html) does __not__ look like an API Reference, it looks like some general documentation. Imagine a new user trying to find a specific function - let's say we want to understand how to create a new array and fill it with a specific value. Let's look for `fill`. No `fill` [here](https://numpy.org/doc/stable/reference/index.html). Now we have to either read the entire page (highly unlikely) or just search. [Boom](https://numpy.org/doc/stable/search.html?q=fill), 30+ results. Wtf? So we go to [another search engine](https://duckduckgo.com/?q=numpy+fill&ia=web). It turns out that we are probably looking for `full`.

Now don't get me wrong, I'm not saying that `full` should not be under _Array creation routines_. This is absolutely accurate and absolutely useless if you want to find something.

[This view](https://numpy.org/doc/stable/reference/routines.array-creation.html) looks a bit more like an API Reference. We have a list of functions and some short descriptions. Sweet! But when wanting to understand what the framework can do probably we would need to use functionality that fits into different documentation sections. For example, we created our array with `full` now we want to split it. Oh, no `split` [here](https://numpy.org/doc/stable/reference/routines.array-creation.html). [Here we go](https://numpy.org/doc/stable/search.html?q=split) [again](https://duckduckgo.com/?q=numpy+split&ia=web)!

I hope you get the point - the users of your framework should not be required to guess or know the taxonomy for the functionality in order to use that functionality. An API Reference should present everything that's available in the framework in a succinct and efficient manner.

### Complex search is probably pointless

Let's face it - you are not Google. No point in reinventing the search. If the search functionality on your page is anything more than a simple JavaScript filter across a few tags then you are probably doing it wrong. This is effort wasted that could have been used more productively. Again, 30+ results of all `fill` methods is absolutely accurate information but useless.

**Comment by melissawm** [comment]
Hello, @teodorkostov  - just to check, would you say having all the reference documentation in one single page would be more helpful? I get what you are saying so I went looking for other documentation and see how they deal with large APIs.

Similar to the NumPy API Reference structure:
- [Python](https://docs.python.org/3/library/index.html)
- [Pandas](https://pandas.pydata.org/docs/reference/index.html)
- [PyTorch](https://pytorch.org/docs/stable/index.html)
- [Django](https://docs.djangoproject.com/en/3.2/ref/)

The one I found most similar to what you are describing (in a single page)
- [scikit-learn](https://scikit-learn.org/stable/modules/classes.html)

I think intuitively the organization helps, but that's because I kind of know where things are. So maybe we could offer both an organized view and a flat view with all the reference docs in a single page.

**Comment by teodorkostov** [comment]
Hello, @melissawm. Yeah, it seems to be some kind of bias in the community. You can add more to that list, e.g. `scipy`. Kudos to the `scikit-learn` team.

The simplicity seems to be something common for JS frameworks though. Look at [D3.js](https://github.com/d3/d3/blob/main/API.md) for example.

I would leave Python out of it. Languages and complex frameworks could require context and examples to make sense of the structures.

`numpy` is a great project. No overcomplicated control structures or type creation. So the documentation can be presented to the user in a straight manner.

**Comment by adeak** [comment]
I can't help but feel that you are conflating some things here.

The way I'm reading your post is that you want the API reference to be a source of learning for new users.

>  Imagine a new user trying to find a specific function - let's say we want to understand how to create a new array and fill it with a specific value. 

That's not what an API reference is for. This is what tutorials are for. [NumPy has many kinds](https://numpy.org/doc/stable/user/index.html). Have you checked if they are appropriate to teach a new user how to do the basics? Maybe not, but this is a tutorial/onboarding issue, not an API reference design issue.

> The only way the this can be used currently is by searching in a third party search engine. 

Actually, the website has its own search bar if you are looking for something. Then again I disagree that using a third-party search engine is inherently a sign of bad design. More often than not the search engine will have a better idea of what you're looking for. Just take your `fill` example: surely you don't expect to explore a new library (with the size of NumPy!) by pressing ctrl+F and looking for random keywords that you _think_ describe what you want to do?

My point is that an API reference should be like a dictionary. You want to know what a word means, you look it up. If it has links to other definitions, you can look those up too (and NumPy is chock full of "see also" notes!). And with this premise it's less crucial what the reference is structured like. If you know what you're looking for, the built-in search bar works fine too, no need for google.

Having a huge wall of descriptions on a single page (and it would be _huge_, because NumPy's API is huge) has issues of its own. In particular, it makes anything other than ctrl+F-based browsing impossible. Imagine trying to find anything using a mobile browser. There are probably also non-trivial accessibility considerations here, but I don't know anything about that to know which design would be preferable there.

**Comment by melissawm** [comment]
> Having a huge wall of descriptions on a single page (and it would be _huge_, because NumPy's API is huge) has issues of its own. In particular, it makes anything other than ctrl+F-based browsing impossible. Imagine trying to find anything using a mobile browser. There are probably also non-trivial accessibility considerations here, but I don't know anything about that to know which design would be preferable there.
 
I agree, but I'd like to understand better... Maybe we're just biased and don't see the value for users that are not already expecting that structure. Then again the complete NumPy docs in pdf format is over 1200 pages long....

**Comment by rkern** [comment]
Due to the history of the documentation (originally sold as a book, then donated to the project; see the [Acknowledgements](https://numpy.org/doc/stable/reference/)), it does have something of a mixed nature. It interleaves bare listing of functions with bits of narrative that would probably be better considered as "User's Guide" or "Reference Manual" documentation than strictly an "API Reference". After all, the original title was [_Guide to NumPy_](https://archive.org/details/NumPyBook), not _NumPy API Reference_.

If all @teodorkostov wants is an additional page that gathers up all of the "walls of descriptions" into a single page, I'd support his volunteering to try that out. We'll have to evaluate how well that can be maintained and evaluate if it actually ends up providing the benefits that are being claimed. I would not go to the effort of entirely restructuring the documentation such that the narrative bits are completely separated from any "walls of descriptions", but having an alternative single-page index of routines could be worth the experiment, if there were a volunteer.

**Comment by teodorkostov** [comment]
Thanks @rkern for keeping an open mind. Essentially, there are already these "walls of descriptions" in the `numpy` docs - [here](https://numpy.org/doc/1.21/reference/routines.array-creation.html). Problem with these is that they are separate pages - the content of [_Array manipulation routines_](https://numpy.org/doc/1.21/reference/routines.array-manipulation.html) is not in the HTML of [_Array creation routines_](https://numpy.org/doc/1.21/reference/routines.array-creation.html). If this was a single page with with a little bit of JS to collapse/expand the sections on the left and to show/hide the content on the middle this would be a solution. Having all the content on one page would make it searchable in the browser. No need to change the detail pages like [this one](https://numpy.org/doc/1.21/reference/generated/numpy.ones.html). Also, it is possible to make the search on the top left display information that is on that page and present it in a nice manner. Such a page should not be 1200 pages long. __There is no need to fundamentally change anything in the documentation. It's just visual changes in order to make the site more user friendly.__

I disagree with @adeak on almost everything. It feels that you have a strong bias and come from the presumption of "If you know what you're looking for". Well, the search and the docs of `numpy` may be useful for core developers. I would disagree that they are made for new users (that actually need to read that stuff). And yes, using Google to find something on a docs page that even has it's own search means a very poor design. However, "My point is that an API reference should be like a dictionary." is spot on. Well, is the API Reference 1.21 like a dictionary right now? When you want to look up a word in a dictionary are you required to traverse a tree of all the characteristics of the language before you are shown a select section of words that may or may not contain the word you are looking for?

**Comment by mattip** [comment]
We discussed this at the latest documentation meeting. Different people want different ways to access information. We would be open to a PR that creates a new page, something like the [scikit learn page](https://scikit-learn.org/stable/modules/classes.html) linked above with the caveats:
- it would have to be generated, not a manually curated, list of API interfaces
- it should only list top-level functions and classes (with their class methods). We could have a similar flat API reference for the sub-namespaces like random, fft, linalg.

It may be extremely long and heavy, which would be suitable for some users and not suitable for others. Perhaps we could prevent mobile users from seeing it with some CSS magic if needed.

**Comment by adeak** [comment]
I assume there'd have to be some preamble explaining the existence of the API dump page. It could be mentioned there that mobile users will probably able to make less use of it. Still, I think it's OK if users have a choice to see it.
